### PR TITLE
Axsuarez/sso protocol

### LINKED ISSUE #711: SSO Implementation: Port from C#
Dotnet PR: https://github.com/microsoft/botbuilder-dotnet/pull/3360

### PR DESCRIPTION
**Pull request by axelsrz**
fixes #711

### CODE REVIEW

File: libraries/botbuilder-core/botbuilder/core/activity_handler.py [hunk 1]
```diff
@@ -344,3 +351,56 @@ class ActivityHandler:
...
+        :param turn_context: The context object for this turn
+        :type turn_context: :class:`botbuilder.core.TurnContext`
+
+        :returns: A task that represents the work queued to execute
+        """
+
+        if (
+            turn_context.activity.name == SignInConstants.verify_state_operation_name
+            or turn_context.activity.name
+            == SignInConstants.token_exchange_operation_name
```
Comment (carlosscastro) [inline libraries/botbuilder-core/botbuilder/core/activity_handler.py]: Why line 370 is not at the end of 369? We all have wide screen monitors, doesn't seem useful to limit to 70 columns

File: libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py [hunk 2]
```diff
@@ -55,25 +60,42 @@
...
+        "bot_url": {"key": "connectionName", "type": "str"},
         "ms_app_id": {"key": "msAppId", "type": "str"},
     }
 
     def __init__(
         self,
         *,
         connection_name: str = None,
-        conversation: ConversationReference = None,
+        conversation=None,
```
Comment (carlosscastro) [inline libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py]: other parameters use spaces around the equals, but your new ones here don't. Please add the spaces for consistency.

File: libraries/botbuilder-core/botbuilder/core/activity_handler.py (hunk 1 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/activity_handler.py]: The python SDK follows a formatting style called black (https://black.readthedocs.io/en/stable/) and checks the formatting in every build.

The formatting is done automatically.

Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/activity_handler.py]: pd. I will try to do it 1 line and check if the formatter doesn't complain

File: libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py (hunk 2 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py]: Yes, this is also part of the formatting, the convention is that if you use typing the format is:

`variable: VariableType = value`

but if you don't use typing the format is:

`variable=value`

Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py]: Formatting aside, I removed the typing accidentally... I should put it back

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py [hunk 3]
```diff
@@ -384,9 +401,78 @@ def __init__(
...
+                    self._settings.oath_app_credentials,
+                    self._settings.connection_name,
+                    context.activity.from_property.id,
+                    TokenExchangeRequest(token=context.activity.value.token),
+                )
+
+                if not token_exchange_response or not token_exchange_response.token:
+                    await context.send_activity(
+                        self._get_token_exchange_invoke_response(
+                            409,
```
Comment (carlosscastro) [inline libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: In other languages we have status code constants (in javascript StatusCodes) so for example you can do `StatusCodes.OK`, etc. Please add that in python if not already existing, or use it if it already exists.

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py [hunk 4]
```diff
@@ -384,9 +401,78 @@ def __init__(
...
+                            "The bot is unable to exchange token. Proceed with regular login.",
+                        )
+                    )
+                else:
+                    await context.send_activity(
+                        self._get_token_exchange_invoke_response(
+                            200, None, context.activity.value.id
```
Comment (carlosscastro) [inline libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: StatusCodes

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py [hunk 5]
```diff
@@ -384,9 +401,78 @@ def __init__(
...
+                        "The bot received an InvokeActivity with a TokenExchangeInvokeRequest containing a"
+                        " ConnectionName that does not match the ConnectionName expected by the bots active"
+                        " OAuthPrompt. Ensure these names match when sending the InvokeActivityInvalid"
+                        " ConnectionName in the TokenExchangeInvokeRequest",
+                    )
+                )
+            elif not getattr(context.adapter, "exchange_token"):
+                await context.send_activity(
+                    self._get_token_exchange_invoke_response(
+                        502,
```
Comment (carlosscastro) [inline libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: In other languages we have status code constants (in javascript StatusCodes) so for example you can do `StatusCodes.OK`, etc. Please add that in python if not already existing, or use it if it already exists.

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py [hunk 6]
```diff
@@ -384,9 +401,78 @@ def __init__(
...
+                        "The bot received an InvokeActivity that is missing a TokenExchangeInvokeRequest value."
+                        " This is required to be sent with the InvokeActivity.",
+                    )
+                )
+            elif (
+                context.activity.value.connection_name != self._settings.connection_name
+            ):
+                await context.send_activity(
+                    self._get_token_exchange_invoke_response(
+                        400,
```
Comment (carlosscastro) [inline libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: In other languages we have status code constants (in javascript StatusCodes) so for example you can do `StatusCodes.OK`, etc. Please add that in python if not already existing, or use it if it already exists.

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py [hunk 7]
```diff
@@ -384,9 +401,78 @@ def __init__(
...
+                context.activity.value = TokenExchangeInvokeRequest().from_dict(
+                    context.activity.value
+                )
+            if not (
+                context.activity.value
+                and self._is_token_exchange_request(context.activity.value)
+            ):
+                await context.send_activity(
+                    self._get_token_exchange_invoke_response(
+                        400,
```
Comment (carlosscastro) [inline libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: In other languages we have status code constants (in javascript StatusCodes) so for example you can do `StatusCodes.OK`, etc. Please add that in python if not already existing, or use it if it already exists.

**Comment by carlosscastro** [comment]
Since you are missing the constants for StatusCodes, this leads me to believe that this is a port from javascript from a few days back, since Swagat added that earlier this week. I'd suggest to, in addition to my comments, review the last commits in the JS PR (including some from last night) and update this PR with those commits.

File: libraries/botbuilder-core/botbuilder/core/activity_handler.py [hunk 8]
```diff
@@ -344,3 +351,56 @@ class ActivityHandler:
             conversation update, message reaction, or event activity, it calls this method.
         """
         return
+
+    async def on_invoke_activity(  # pylint: disable=unused-argument
+        self, turn_context: TurnContext
+    ):
```
Comment (swagatmishra2007) [inline libraries/botbuilder-core/botbuilder/core/activity_handler.py]: john is makign some changes to invoke to make it more consistent with teams. you might want to port that work to python or have him port it (he is doing it for c# and js).
i am not sure what the work is

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py (hunk 6 above)
Comment (tracyboehrer) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: Corrected

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py (hunk 7 above)
Comment (tracyboehrer) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: Corrected

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py (hunk 5 above)
Comment (tracyboehrer) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: Corrected

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py (hunk 4 above)
Comment (tracyboehrer) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: Corrected

File: libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py (hunk 3 above)
Comment (tracyboehrer) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/prompts/oauth_prompt.py]: Corrected

File: libraries/botbuilder-core/botbuilder/core/activity_handler.py [hunk 9]
```diff
@@ -344,3 +352,56 @@ class ActivityHandler:
...
+        By default, this method does nothing.
+
+        :param turn_context: The context object for this turn
+        :type turn_context: :class:`botbuilder.core.TurnContext`
+
+        :returns: A task that represents the work queued to execute
+        """
+        return
+
+    async def on_invoke(  # pylint: disable=unused-argument
```
Comment (johnataylor) [inline libraries/botbuilder-core/botbuilder/core/activity_handler.py]: We don't want this one -  the model is flatter for invoke - this looks like a copy of the event-activity pattern.

Specifically we want an overridable onInvokeActivity and it should return an invoke-response then we want code in the onTurn that takes that invoke-response, creates an invoke-response-activity and sends it. The code inline in onTurn should do this send conditionally - if the override has already done a send then don't do one here - refer to the code in C# (ActivityHandler) and JS (ActivityHandlerBase).

Next having introduced this onInvokeActivity method in ActivityHandler we can remove the now duplicate code in TeamsActivityHandler.

So teh change is basically to move the code up the class hieerarchy from TeamsActivityHandler into ActivityHandler

Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/activity_handler.py]: I just pushed those changes.

File: libraries/botbuilder-core/botbuilder/core/activity_handler.py [hunk 10]
```diff
@@ -344,3 +364,58 @@ class ActivityHandler:
...
+            if (
+                turn_context.activity.name
+                == SignInConstants.verify_state_operation_name
+                or turn_context.activity.name
+                == SignInConstants.token_exchange_operation_name
+            ):
+                await self.on_sign_in_invoke(turn_context)
+                return self._create_invoke_response()
+
+            raise _InvokeResponseException(HTTPStatus.NOT_IMPLEMENTED)
```
Comment (johnataylor) [inline libraries/botbuilder-core/botbuilder/core/activity_handler.py]: you have 502 behind the constant **HTTPStatus**.NOT_IMPLEMENTED but 200 behind **StatusCodes**.OK

that reads a little wonky - is there are reason for this?

File: libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py [hunk 11]
```diff
@@ -1109,7 +1234,7 @@ def key_for_connector_client(service_url: str, app_id: str, scope: str):
             self._is_emulating_oauth_cards = True
 
         app_id = self.__get_app_id(context)
-        scope = context.turn_state[BotAdapter.BOT_OAUTH_SCOPE_KEY]
+        scope = context.turn_state.get(BotAdapter.BOT_OAUTH_SCOPE_KEY)
```
Comment (johnataylor) [inline libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py]: what is the difference between using [] and get() ?

File: libraries/botbuilder-core/botbuilder/core/status_codes.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    OK = 200
+    SUCCESSFUL_LIMIT = 299
+    BAD_REQUEST = 400
+    UNAUTHORIZED = 401
+    NOT_FOUND = 404
+    METHOD_NOT_ALLOWED = 405
+    CONFLICT = 409
+    UPGRADE_REQUIRED = 426
+    INTERNAL_SERVER_ERROR = 500
+    NOT_IMPLEMENTED = 501
```
Comment (johnataylor) [inline libraries/botbuilder-core/botbuilder/core/status_codes.py]: so you do have NOT_IMPLEMENTED - why not use it? (or alternatively just use the HTTPStatus everywhere.)

File: libraries/botbuilder-core/botbuilder/core/activity_handler.py (hunk 8 above)
Comment (johnataylor) [reply libraries/botbuilder-core/botbuilder/core/activity_handler.py]: see other comments.
do we need to refer to HTTPStatus and StatusCodes?

File: libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py (hunk 11 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/bot_framework_adapter.py]: get does not raises KeyError if element does not exists, just returns None

File: libraries/botbuilder-core/botbuilder/core/status_codes.py (hunk 12 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/status_codes.py]: Yes, this was due to me doing a literal porting of js PR... I might get rid of StatusCodes

**Pull request merged**